packages/next/src/server/web/sandbox/context.ts

```diff
@@ -216,7 +217,15 @@ Learn More: https://nextjs.org/docs/messages/edge-dynamic-code-evaluation`),
           return result
         }
 
-      const __fetch = context.fetch
+      const __fetch__ = context.fetch
+      const __fetch: typeof __fetch__ = async (...args) => {
+        try {
+          return await __fetch__(...args)
+        } catch (err: any) {
+          Error.captureStackTrace(err, context.fetch)
+          throw err
+        }
+      }
```

Comment:
This isn't enough to handle the full issue: https://gist.github.com/jridgewell/4e6b9ceb954fffa9013ad8a2f9214c90#file-3-async-await-broken-js

You need to close over a new `Error` instance (let's call it `source`), and reattach `source`'s stack trace to the thrown error to carry over the full call stack. https://gist.github.com/jridgewell/4e6b9ceb954fffa9013ad8a2f9214c90#file-2-node-fetch-fix-js-L37

Reply:
Hey @jridgewell! Thanks for the review (and teaching me about `Error.captureStackTrace` <picture data-single-emoji=":just-right2:" title=":just-right2:"><img class="emoji" width="20" height="auto" src="https://single-emoji.vercel.app/api/emoji/eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..l65rJhUHakIqjZsk.KuuKQqdM5TxutCYGzxDXOgAn-VGU7Q6pdSkh643Lr9knkBjnGZfOAN1eVg18PB7EjjJNwxhIg9agXxXqItkD2HjDpkT2r2pKIDDojflj4Q7O2KGvaJItB167fF4k.4rDvPmgMARhL-md6_8xU-A" alt=":just-right2:" align="absmiddle"></picture> ) and sorry for postponing that for a while. What would this change do in terms of my expectations in my tests? As they seem fine and I want to make sure I cover all the cases (kind of a TDD approach) because stack traces are tricky <picture data-single-emoji=":blob-angel:" title=":blob-angel:"><img class="emoji" width="20" height="auto" src="https://single-emoji.vercel.app/api/emoji/eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..U-iiNidxSoBeTnR-.vzOwdseXeoALU8nLDjuGYlpNtdxCUzCfOPRnYC2WvFCGQrvMN0UHm1F-YVP5zHeBuJKHAQ00annn0q9Fm7lYS0zejI7-2wIjnv5remRp6BOXq4HpxbJlX3j6V3o.GB3hMNNa7CpGThBVep7ESA" alt=":blob-angel:" align="absmiddle"></picture> 

Seems that using async/await is necessary in any way to track the stack correctly. Floating promises and/or returning a `Promise` without awaiting seem to break stack traces no matter what.

But if there's an edge case I don't cover, I'd like to add it to the tests so it will be explicitly tested and your suggestion would be the fix <picture data-single-emoji=":just-right2:" title=":just-right2:"><img class="emoji" width="20" height="auto" src="https://single-emoji.vercel.app/api/emoji/eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..l65rJhUHakIqjZsk.KuuKQqdM5TxutCYGzxDXOgAn-VGU7Q6pdSkh643Lr9knkBjnGZfOAN1eVg18PB7EjjJNwxhIg9agXxXqItkD2HjDpkT2r2pKIDDojflj4Q7O2KGvaJItB167fF4k.4rDvPmgMARhL-md6_8xU-A" alt=":just-right2:" align="absmiddle"></picture> <picture data-single-emoji=":yesss:" title=":yesss:"><img class="emoji" width="20" height="auto" src="https://single-emoji.vercel.app/api/emoji/eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..XBi0_QfrklBVnRua.rQvvqGxyCQ3qfmQOwk3I2duGeRv-ji17MWRmTKwExooI8LcLyM9fXzLEnTumebTp8DF8Y5DcmqaDiS4cw7D83u1tzqqCWRfjjqB-Uj3-UcFYY8bukz71.VU1nc-ioa8BWpAOA7KAXdA" alt=":yesss:" align="absmiddle"></picture> 

so can you please help me with this one?

Reply:
I don't know how to run the test suite, so I'm making assumptions:

> Seems that using async/await is necessary in any way to track the stack correctly. Floating promises and/or returning a `Promise` without awaiting seem to break stack traces no matter what.

This is the core problem, and I think it's why your tests are passing. You're using `await`s all the way down your chain, which the user is not always guaranteed to do. If the user has any promise usage, then we've lost the context from before that promise.

To illustrate this, I've made https://github.com/jridgewell/stack-trace-async-event, where "current" is the current production behavior, "PR" is this PR's behavior, and "Final" is the behavior I'm trying to propose. With this PR, we're broken in promise code, and fixed in await code.

The cases that match this PR are 4, 5, 6. The difference between them is how whether we `await` the promise in the current function. If the user only uses `await`, and does it for the entire call stack, we're fine (and that's what's tested). But if a user uses a Promise in the stack, then that promise frame will be ommitted (case 5). And if the user only uses promise, then all frames are omitted (case 4).

With my proposed `source` error stack capturing, we fix this. It won't matter if the user used promises or await, because we won't depend on V8's async stack stitching (which only works for `await`). Instead, we'll have access to the stack at the moment the `fetch` was called (lazily, too, via the `source` error's stack). That will capture the async and sync calls stack that got us to this point, which will include promises.



> But if there's an edge case I don't cover, I'd like to add it to the tests so it will be explicitly tested and your suggestion would be the fix

For test cases, we need to throw in some non-awaited promises. See https://github.com/jridgewell/stack-trace-async-event/blob/main/7-promises-capture-source.js and https://github.com/jridgewell/stack-trace-async-event/blob/main/8-some-async-awaits-capture-source.js to complement you're already existing https://github.com/jridgewell/stack-trace-async-event/blob/main/9-all-async-awaits-capture-source.js.